aea/helpers/async_utils.py

```diff
@@ -0,0 +1,324 @@
+    include_cancelled: bool = False,
+    loop: Optional[AbstractEventLoop] = None,
+) -> List[Exception]:
+    """
+    Wait first task completed or exception raised and cancel other tasks.
+
+    :param tasks: list of tasks to run and wait.
+
+    :return: list of exceptions raised.
+    """
+    exceptions: List[Exception] = []
+
+    _, pending = await asyncio.wait(tasks, return_when=FIRST_COMPLETED)
+
+    for task in pending:
+        task.cancel()
+
+    completed, pending = await asyncio.wait(tasks, return_when=ALL_COMPLETED)
+
+    assert not pending
+
+    for task in completed:
+        if not include_cancelled and task.cancelled():
+            continue
+        exc = task.exception()
+        if exc:
+            exceptions.append(cast(Exception, exc))
+
+    return exceptions
+
+
+class AnotherThreadTask:
+    """
+    Schedule a task to run on the loop in another thread.
+
+    Provides better cancel behaviour: on cancel it will wait till cancelled completely.
+    """
+
+    def __init__(self, coro: Coroutine, loop: AbstractEventLoop) -> None:
+        """
+        Init the task.
+
+        :param coro: coroutine to schedule
+        :param loop: an event loop to schedule on.
+        """
+        self._task = loop.create_task(coro)
+        self._loop = loop
+        self._future = asyncio.run_coroutine_threadsafe(self._get_task_result(), loop)
+
+    async def _get_task_result(self) -> Any:
+        """Get task result, should be run in target loop."""
+        return await self._task
+
+    def result(self, timeout: Optional[float] = None) -> Any:
+        """
+        Wait for coroutine execution result.
+
+        :param timeout: optional timeout to wait in seconds.
+        """
+        return self._future.result(timeout)
+
+    def cancel(self) -> None:
+        """Cancel coroutine task execution in a target loop."""
+        self._loop.call_soon_threadsafe(self._task.cancel)
+
+    def future_cancel(self) -> None:
+        """
+        Cancel task waiting future.
+
+        In this case future result will raise CanclledError not waiting for real task exit.
+        """
+        self._future.cancel()
+
+    def done(self) -> bool:
+        """Check task is done."""
+        return self._future.done()
+
+
+class ThreadedAsyncRunner(Thread):
+    """Util to run thread with event loop and execute coroutines inside."""
+
+    def __init__(self, loop=None) -> None:
+        """
+        Init threaded runner.
+
+        :param loop: optional event loop. is it's running loop, threaded runner will use it.
+        """
+        self._loop = loop or asyncio.new_event_loop()
+        assert not self._loop.is_closed()
+        super().__init__(daemon=True)
+
+    def start(self) -> None:
+        """Start event loop in dedicated thread."""
+        if self.is_alive() or self._loop.is_running():
+            return
+        super().start()
+        self.call(asyncio.sleep(0.001)).result(1)  # type: ignore # to ensure loop is started
```

Comment:
my IDE's linter gives: 

![image](https://user-images.githubusercontent.com/9467617/83245436-04c44680-a1a1-11ea-8ec7-98fd4a8db71e.png)

Reply:
same for me, this is why it ignored in comments, cause code is valid and works as expected

the reason can be: different implementation in python3.8 :
```
async def sleep(delay, result=None, *, loop=None):
    """Coroutine that completes after a given time (in seconds)."""
    if delay <= 0:
        await __sleep0()
        return result
```

Reply:
updated type hints to Awaitable. should help